### ISSUE TITLE
fatJar contains optional dependencies

### ISSUE THREAD

**Issue by cbeust** [cbeust/kobalt#316]
Need to understand Aether better so I can exclude optional dependencies from fat jars.

The starting point is `DependencyManager#transitiveClosure`.

**Comment by cbeust** [comment]
Still trying to debug this. I'm deep in the bowels of Aether (`DefaultDependencyNode`) and I see incorrect things such as:

```
this = {DefaultDependencyNode@3886} "com.squareup.retrofit2:retrofit:jar:2.1.0 (compile)
dependency = {Dependency@3910} "com.google.android:android:jar:4.1.1.4 (compile)"
 artifact = {DefaultArtifact@3911} "com.google.android:android:jar:4.1.1.4"
 scope = "compile"
 optional = {Boolean@3923} "false"
"
```

This is Aether traversing the dependencies of Retrofit2 and (incorrectly) counting Android as a dependency that is **not** optional (even though the `.pom` says it is:

```
    <dependency>
      <groupId>com.google.android</groupId>
      <artifactId>android</artifactId>
      <optional>true</optional>
```

Bit mystified right now.

**Comment by cbeust** [comment]
/cc @pabl0rg

Ok I don't even know where to begin with this. It drove me crazy.

I dove deep inside Aether, found where the problem was occurring but obviously, no real idea why it was implemented this way. So I went back to the `aether-demo` repo, ran all the tests and example and they all passed. Copy/pasted the example in the Kobalt code base, ran it and... the optional dependencies are still showing up.

I aligned both codes to be exactly similar, same result: works in `demo-examples`, fails in `kobalt`.

Then I took a look at the versions and I noticed that Kobalt was using the latest Aether implementation (1.1.0) while `aether-demo` is using an `1.0` snapshot. I updated the Kobalt dependencies to use that same snapshot and... boom, it works.

I'm so mad right now.

I then built your example and verified that the Android dependencies are no longer included (and the size of the jar file is now just under 10M).

Going to do some clean up and publish that new version soon.
